/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
syntax = "proto3";

package profiler.proto;
option java_package = "com.android.tools.profiler.proto";
option java_outer_classname = "Transport";

import "common.proto";

service TransportService {
  // Gets current timestamp that used in the device's data.
  // A client may use this API to relate data' timestamps to other
  // clocks such as the desktop's calendar time.
  rpc GetCurrentTime(TimeRequest) returns (TimeResponse) {
  }

  // Gets version.
  rpc GetVersion(VersionRequest) returns (VersionResponse) {
  }

  // Query for cached byte data associated with some ID. Other services may
  // return an ID which can be used to fetch data using this central cache.
  rpc GetBytes(BytesRequest) returns (BytesResponse) {
  }

  // Gets the currently connected devices.
  // TODO: replace with command in the new pipeline.
  // TODO: expand with timing information for historical requests.
  rpc GetDevices(GetDevicesRequest) returns (GetDevicesResponse) {
  }

  // Gets the running processes of a given device
  // TODO: replace with command in the new pipeline.
  rpc GetProcesses(GetProcessesRequest) returns (GetProcessesResponse) {
  }

  // Query for whether the agent is alive.
  // TODO: replace with command in the new pipeline.
  rpc GetAgentStatus(AgentStatusRequest) returns (AgentData) {
  }

  // Makes all necessary configurations needed for attaching agent on startup.
  // Returns |agent_args| to be used with the application launch command,
  // e.g "am start --attach-agent |agent_args|".
  // TODO: replace with command in the new pipeline.
  rpc ConfigureStartupAgent(ConfigureStartupAgentRequest)
      returns (ConfigureStartupAgentResponse) {
  }

  // Executes a |Command| on the daemon.
  rpc Execute(ExecuteRequest) returns (ExecuteResponse) {
  }

  // Streams events as they are being generated by the app or the perfd daemon.
  // This API is most useful for getting all profiling data regardless of any
  // filtering, for example, from a device.
  rpc GetEvents(GetEventsRequest) returns (stream Event) {
  }

  // Gets all the event groups (events that share the same group_id) that
  // intersect the range [|from_timestamp|, |to_timestamp|]. If the paramter
  // |end| is used, an event_group will be considered "open" until such an event
  // shows up. For example if we have a "SESSION_STARTED" at time 1,  and the
  // range requested is [2,3], and the |end| field is set to SESSEND_SESSION",  then
  // the event group will be considered opened, and this intersecting the range.
  rpc GetEventGroups(GetEventGroupsRequest) returns (GetEventGroupsResponse) {
  }
}

message TimeRequest {
  int64 stream_id = 1;
}

message TimeResponse {
  // What's used as timestamps in all profiler data (unless otherwise
  // specified). (a.k.a: Timestamp from clock_gettime(CLOCK_MONOTONIC, ...), ns
  // precision.)
  int64 timestamp_ns = 1;

  // The number of microseconds since the Epoch.
  // (a.k.a: Timestamp from gettimeofday, us precision.)
  int64 epoch_timestamp_us = 2;
}

message VersionRequest {
  int64 stream_id = 1;
}

message VersionResponse {
  string version = 1;
}

message BytesRequest {
  int64 stream_id = 1;
  // ID for fetching contents from a cache. The value will always be safe to use
  // as a filename, if you want to cache the contents locally to disk for
  // example.
  string id = 2;
}

message BytesResponse {
  // Byte contents from a cache, or "" if no data was found associated with the
  // cache key used to fetch it, or if the data was removed after some timeout.
  bytes contents = 1;
}

message GetDevicesRequest {
}

message GetDevicesResponse {
  repeated Device device = 1;
}

message GetProcessesRequest {
  int64 device_id = 1;
}

message GetProcessesResponse {
  repeated Process process = 2;
}

message AgentStatusRequest {
  int64 device_id = 1;
  int32 pid = 2;
}

message ConfigureStartupAgentRequest {
  int64 stream_id = 1;
  string app_package_name = 2;
  string agent_lib_file_name = 3;
}

message ConfigureStartupAgentResponse {
  string agent_args = 1;
}

message ExecuteRequest {
  Command command = 1;
}

message ExecuteResponse {
}

message GetEventsRequest {
}

// Every field in this message serves the purpose of filtering if it is set.
// When multiple fields are set, all filtering conditions must be met so an
// event group can be returned.
//
// The default value of a proto type (e.g., 0 for int64) is not considered
// valid. For instance, 0 is not a valid session ID.
message GetEventGroupsRequest {
  // An event is returned if |stream_id| matches.
  int64 stream_id = 1;
  // An event is returned if |pid| matches.
  int32 pid = 2;
  // An event is returned if |kind| matches.
  Event.Kind kind = 3;
    // An event is returned if |group_id| matches.
  int64 group_id = 4;
  // When |from_timestamp| and/or |to_timestamp| are set, an event group is
  // returned including the events during with the range, also including the
  // last event before |from_timestamp| and the first one after |to_timestamp|
  // in each group.
  int64 from_timestamp = 5;  // inclusive
  int64 to_timestamp = 6;    // inclusive
}

message GetEventGroupsResponse {
  repeated EventGroup groups = 1;
}

// An event group is a collection of events that share the same group_id.
// The events in a group should have the same |session_id| and |kind|.
message EventGroup {
  int64 group_id = 1;
  repeated Event events = 2;
}

///////////////////////////////////////////////////////////////////////
// BEGIN Command proto definitions

message Command {
  enum CommandType {
    UNSPECIFIED = 0;
    BEGIN_SESSION = 1;
    END_SESSION = 2;
  }

  // A stream is an entity that provides data, such as a device or an imported
  // file.
  int64 stream_id = 1;  // If ommited the command is applied to all streams
  CommandType type = 2;
  oneof union {
    BeginSession begin_session = 3;
    EndSession end_session = 4;
  }
}

message BeginSession {
  message JvmtiConfig {
    // True if an jvmti agent should be attached, false otherwise.
    bool attach_agent = 1;
    // TODO(b/65458869): Remove when moving process discovery to perfd.
    string agent_lib_file_name = 2;
    // True if live allocation is used, false otherwise.
    bool live_allocation_enabled = 3;
  }

  int32 pid = 1;
  JvmtiConfig jvmti_config = 2;
  int64 request_time_epoch_ms = 3;
  string session_name = 4;
}

message EndSession {
  int64 session_id = 1;
}

// END Command proto definitions
///////////////////////////////////////////////////////////////////////